src/Microsoft.AspNet.OData.Shared/Routing/Conventions/RequestMethodExtensions.cs

```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNet.OData.Interfaces;
+
+namespace Microsoft.AspNet.OData.Routing.Conventions
+{
+    internal static class RequestMethodExtensions
+    {
+        /// <summary>
+        /// Returns the request method and in the case of Options request it returns the Access-Control-Request-Method present in the
+        /// preflight request for the request method that will be used for the actual request
+        /// </summary>
+        /// <param name="request"></param>
+        /// <returns></returns>
+        internal static ODataRequestMethod GetRequestMethodOrPreflightMethod(this IWebApiRequestMessage request)
```

Comment:
>GetRequestMethodOrPreflightMethod [](start = 43, length = 33)

What if the developer already has explicit logic to implement OPTIONS? Would this override a custom implementation? Would they have to rely on ordering to check for their implementation first?

Reply:
I believe this won't block their current usecase. This fix exposes the endpoint metadata for the cors middleware to use.

Reply:
they can also override the behaviour easily by extending the value transformer or providing a custom one.

Reply:
I had explicit logic to implement OPTIONS and this blocked my use case.

@marabooy could you show me how to follow your suggestion to customize/override the transformer? I could use a hand.

Reply:
@GrayPockets do you mean the routing or response transformers?

Reply:
@GrayPockets You could use the following.

Override the TransformAsync method in [the ODataEndpointRouteValueTransformer  class.](https://github.com/OData/WebApi/blob/2ccbefd65e19e13f015f653f3c446a6a19f08406/src/Microsoft.AspNetCore.OData/Extensions/Endpoint/ODataEndpointRouteValueTransformer.cs#L130) The default method sets an endpoint so that the default cors middleware can read the annotations on the method or controller.  Do let me know if you do experience issues with this approach, I would be happy to create a fix to make it easier for users to override this.

Reply:
> Override the TransformAsync method in the ODataEndpointRouteValueTransformer class. 

@marabooy perhaps I am dense, I do not see how I can do this.
Attempting to create a class to inherit from ODataEndpointRouteValueTransformer in the project of my API gives me tons of accessibility/protection errors.

![image](https://user-images.githubusercontent.com/17136682/139744378-a4b797c6-aac5-4b5d-a017-2707869fb245.png)

I'm glad that the CORS updates occurred, but I have custom [HttpOptions] methods on my controllers, and this change causes them to throw 404 errors. This allows me to have fine-grain control of the Option header responses to each controller. I'm also using Microsoft.AspNetCore.OData.Versioning, which complicates the middleware layer on top of it.

As a work-around for today, I downloaded the source for this project to replace the NuGet package, and commented out lines 130-140.

If there was setting to disable these lines, I would be greatly appreciative. I may not be able to grok how to make that change on my own.

Reply:
Thanks for the update, Let me create an issue that can be followed to allow extensibility of this. Thanks for the update.

Reply:
Awesome!